### PR TITLE
octopus: mgr/cephadm: pass --container-init to cephadm if specified

### PR DESCRIPTION
**Pull request by tserong**
This reverts a small part of 442904bf2da, in order to work correctly
with 82d30afc413.  The problem is that the latter commit sets
CONTAINER_INIT=False, which means we need to explicitly pass
--container-init to cephadm if we want that behaviour turned on.

Fixes: https://tracker.ceph.com/issues/52064
Signed-off-by: Tim Serong <tserong@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -1123,6 +1123,9 @@ def _run_cephadm(self,
             if not no_fsid:
                 final_args += ['--fsid', self._cluster_fsid]
 
+            if self.container_init:
+                final_args += ['--container-init']
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:1127]: a few lines above we have 

```
1116            if not self.container_init:
1117                final_args += ['--no-container-init']
```
Are those lines correct in octous?

Comment (tserong) [reply src/pybind/mgr/cephadm/module.py:1127]: Yeah -- 442904b adds --no-container-init to cephadm, and I think it's fine to keep both these flags here, as it makes it the cephadm mgr module explicitly specify one or the other, regardless of what the internal cephadm default might be

Comment (smithfarm) [reply src/pybind/mgr/cephadm/module.py:1127]: Would it make sense to combine the two `if` statements into one like so:

```
if self.container_init:
    final_args += ['--container-init']
else:
    final_args += ['--no-container-init']
```

?

**Review by sebastian-philipp — APPROVED** [review]

File: src/pybind/mgr/cephadm/module.py (hunk 1 above)
Comment (tserong) [reply src/pybind/mgr/cephadm/module.py:1127]: I'd like to, but due to the way those parameters are parsed, I don't think I can. `--no-container-init` is accepted at the top level (e.g.: `cephadm --no-container-init [command] ...`) whereas `--container-init` is only accepted inside individual commands (`cephadm adopt --container-init`, `cephadm bootstrap --container-init`, `cephadm deploy --container-init`), so I think it has to remain structured as it is now, i.e.:

```
            if not self.container_init:
                final_args += ['--no-container-init']

            # subcommand
            final_args.append(command)

            # subcommand args
            if not no_fsid:
                final_args += ['--fsid', self._cluster_fsid]

            if self.container_init:
                final_args += ['--container-init']
```

**Review by smithfarm — APPROVED** [review]

**Comment by smithfarm** [comment]
@sebastian-philipp And this problem has been fixed in pacific, right? It seems that in pacific `--container-init` defaults to True:

```
        Option(
            'container_init',
            type='bool',
            default=True,
            desc='Run podman/docker with `--init`'
        ),  
```

**Comment by smithfarm** [comment]
@tserong Are you sure this PR is complete? I see that dropped the `--container-init` option from `cephadm bootstrap`, but this PR doesn't revert that part.

The reason I ask is that, when I apply this PR on top of v15.2.14, `cephadm bootstrap` does not recognize the `--container-init` option (and `cephadm bootstrap --help` does not mention it).

**Comment by tserong** [comment]
I thought `--container-init` was still there, just that the help text had been suppressed? Will check tomorrow.

**Comment by smithfarm** [comment]
> I thought `--container-init` was still there, just that the help text had been suppressed? Will check tomorrow.

AFAICT https://github.com/ceph/ceph/commit/442904bf2da3b83c14d262a5673b81476a537364 dropped it, and this commit doesn't reinstate it.

**Comment by smithfarm** [comment]
@tserong It might be useful to look at https://github.com/ceph/ceph/pull/39914 in its entirety... that's the PR that was supposed to enable `--init` by default everywhere, but apparently didn't quite manage to do so.

**Comment by tserong** [comment]
@smithfarm do you mind trying this updated version?

**Review by smithfarm — APPROVED** [review]
@tserong Works like a charm! Thanks!

**Comment by smithfarm** [comment]
jenkins test make check

**Comment by sebastian-philipp** [comment]
perfect. Thanks for the fix!

File: src/pybind/mgr/cephadm/module.py (hunk 1 above)
Comment (mgfritch) [reply src/pybind/mgr/cephadm/module.py:1127]: yeah, `--container-init` was only accepted for the bootstrap, deploy and adopt commands, whereas the `--no-container-init` param is global to all cephadm commands ...

**Review by mgfritch — APPROVED** [review]
oops, thanks @tserong!

**Pull request merged**